src/sig/sig.c

```diff
@@ -1,79 +1,120 @@
-		if (OQS_SIG_picnic_get(s, algid) != OQS_SUCCESS) {
-			OQS_MEM_insecure_free(s);
-			return NULL;
-		}
-		break;
+OQS_SIG *OQS_SIG_new(const char *method_name) {
+	if (0 == strcasecmp(method_name, OQS_SIG_alg_default)) {
+		return OQS_SIG_new(OQS_SIG_DEFAULT);
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_qTESLA_I)) {
+#ifdef OQS_ENABLE_SIG_qTESLA_I
+		return OQS_SIG_qTESLA_I_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_qTESLA_III_size)) {
+#ifdef OQS_ENABLE_SIG_qTESLA_III_size
+		return OQS_SIG_qTESLA_III_size_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_qTESLA_III_speed)) {
+#ifdef OQS_ENABLE_SIG_qTESLA_III_speed
+		return OQS_SIG_qTESLA_III_speed_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L1_FS)) {
+#ifdef OQS_ENABLE_SIG_picnic_L1_FS
+		return OQS_SIG_picnic_L1_FS_new();
+#else
+		return NULL;
 #endif
-#ifdef ENABLE_SIG_QTESLA
-	case OQS_SIG_qTESLA_I:
-	case OQS_SIG_qTESLA_III_speed:
-	case OQS_SIG_qTESLA_III_size:
-		if (OQS_SIG_qTESLA_get(s, algid) != OQS_SUCCESS) {
-			OQS_MEM_insecure_free(s);
-			return NULL;
-		}
-		break;
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L1_UR)) {
+#ifdef OQS_ENABLE_SIG_picnic_L1_UR
+		return OQS_SIG_picnic_L1_UR_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L3_FS)) {
+#ifdef OQS_ENABLE_SIG_picnic_L3_FS
+		return OQS_SIG_picnic_L3_FS_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L3_UR)) {
+#ifdef OQS_ENABLE_SIG_picnic_L3_UR
+		return OQS_SIG_picnic_L3_UR_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L5_FS)) {
+#ifdef OQS_ENABLE_SIG_picnic_L5_FS
+		return OQS_SIG_picnic_L5_FS_new();
+#else
+		return NULL;
+#endif
+	} else if (0 == strcasecmp(method_name, OQS_SIG_alg_picnic_L5_UR)) {
+#ifdef OQS_ENABLE_SIG_picnic_L5_UR
+		return OQS_SIG_picnic_L5_UR_new();
+#else
+		return NULL;
 #endif
-	default:
-		OQS_MEM_insecure_free(s);
+		// EDIT-WHEN-ADDING-SIG
+	} else {
 		return NULL;
 	}
-
-	return s;
 }
 
-OQS_API OQS_STATUS OQS_SIG_keygen(const OQS_SIG *s, uint8_t *priv, uint8_t *pub) {
-	if (s == NULL) {
+OQS_STATUS OQS_SIG_keypair(const OQS_SIG *sig, uint8_t *public_key, uint8_t *secret_key) {
+	if (sig == NULL || sig->keypair(public_key, secret_key) != 0) {
 		return OQS_ERROR;
 	} else {
-		return s->keygen(s, priv, pub);
+		return OQS_SUCCESS;
 	}
 }
 
-OQS_API OQS_STATUS OQS_SIG_sign(const OQS_SIG *s, const uint8_t *priv, const uint8_t *msg, const size_t msg_len, uint8_t *sig, size_t *sig_len) {
-	if (s == NULL) {
+OQS_STATUS OQS_SIG_sign(const OQS_SIG *sig, uint8_t *signature, size_t *signature_len, const uint8_t *message, size_t message_len, const uint8_t *secret_key) {
+	if (sig == NULL || sig->sign(signature, signature_len, message, message_len, secret_key) != 0) {
```

Comment:
Why are you comparing the output of `sig->sign` with 0 as opposed to `OQS_SUCCESS`?

Reply:
Copy-pasted from nist-branch; fixed here; should fix it there too...

Reply:
Also created PR #393 on nist-branch to fix this